src/awkward1/operations/structure.py

```diff
@@ -432,42 +433,54 @@ def with_field(base, what, where=None, highlevel=True):
     #ak.with_field, so performance is not a factor in choosing one over the
     other.)
     """
-    base = awkward1.operations.convert.to_layout(
-        base, allow_record=True, allow_other=False
-    )
-    if base.numfields < 0:
-        raise ValueError("no tuples or records in array; cannot add a new " "field")
 
-    what = awkward1.operations.convert.to_layout(
-        what, allow_record=True, allow_other=True
-    )
+    if isinstance(where, (list, tuple)) and len(where) > 1:
+        return with_field(
+            base,
+            with_field(base[where[0]], what, where=where[1:], highlevel=highlevel),
+            where=where[0],
+            highlevel=highlevel
+        )
+    else:
+        if isinstance(where, (list, tuple)):
+            where = where[0]
 
-    keys = base.keys()
-    if where in base.keys():
-        keys.remove(where)
-        base = base[keys]
+        base = awkward1.operations.convert.to_layout(
+            base, allow_record=True, allow_other=False
+        )
+        if base.numfields < 0:
+            raise ValueError("no tuples or records in array; cannot add a new " "field")
 
-    def getfunction(inputs, depth):
-        base, what = inputs
-        if isinstance(base, awkward1.layout.RecordArray):
-            if not isinstance(what, awkward1.layout.Content):
-                what = awkward1.layout.NumpyArray(
-                    numpy.lib.stride_tricks.as_strided(
-                        [what], shape=(len(base),), strides=(0,)
+        what = awkward1.operations.convert.to_layout(
+            what, allow_record=True, allow_other=True
+        )
+
+        keys = base.keys()
+        if where in base.keys():
+            keys.remove(where)
+            base = base[keys]
+
+        def getfunction(inputs, depth):
+            base, what = inputs
+            if isinstance(base, awkward1.layout.RecordArray):
+                if not isinstance(what, awkward1.layout.Content):
+                    what = awkward1.layout.NumpyArray(
+                        numpy.lib.stride_tricks.as_strided(
+                            [what], shape=(len(base),), strides=(0,)
+                        )
                     )
-                )
-            return lambda: (base.setitem_field(where, what),)
-        else:
-            return None
+                return lambda: (base.setitem_field(where, what),)
```

Comment:
It looks like this is the only line involved in replacing an existing field with a new field (your "replacing `"a"` with a new `"a"` example that dropped the original array to zero length). If so, then I guess that switch must be in the C++. And yet, I don't see it in [RecordArray::setitem_field](https://github.com/scikit-hep/awkward-1.0/blob/c11597ffaa58a46a05bee2cb36985e6e545188df/src/libawkward/array/RecordArray.cpp#L390-L417).

So it may be that I just never thought of that case, and we're seeing some strange side-effect. The layout nodes' members, such as `base.setitem_field` here, are considered internal, so they don't have to have a great API. A public-facing function like ak.with_field does, however. The case of replacing the only existing field could be handled by some Python code here so that that case never reaches `base.setitem_field`. (The `Content::setitem_field` methods _only_ exist for ak.with_field, so if that case never reaches them here, it's never relevant.)

Lines 458-461 above detect a situation in which you're replacing a field that already exists. The case you want to protect against is when the field you're replacing is the only field, so that's when `len(keys) == 0` after the `keys.remove(where)`.

In that case, there's no reason to go through `base.setitem_field` because

```python
awkward1.layout.RecordArray([what], [where])
```

is already the answer; it wraps a layout node `what` in records with a single field (str) `where`. Inside of the implementation of an operation, it's better to use the low-level constructors than `ak.zip({where: what}, depth_limit=1)` (because the latter returns a high-level ak.Array that would need to be unwrapped to return the layout node).